packages/client-sdk-nodejs/src/internal/grpc/client-timeout-interceptor.ts

```diff
@@ -1,14 +1,123 @@
+      overallRequestTimeoutMs,
+      responseDataReceivedTimeoutMs
+    ).createTimeoutInterceptor();
+  }
+  return new BasicTimeoutInterceptor(
+    overallRequestTimeoutMs
+  ).createTimeoutInterceptor();
 };
+
+export class BasicTimeoutInterceptor {
+  private readonly overallRequestTimeoutMs: number;
+
+  constructor(overallRequestTimeoutMs: number) {
+    this.overallRequestTimeoutMs = overallRequestTimeoutMs;
+  }
+
+  public createTimeoutInterceptor(): Interceptor {
+    return (options, nextCall) => {
+      if (!options.deadline) {
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.overallRequestTimeoutMs
+        );
+        options.deadline = deadline;
+      }
+      return new InterceptingCall(nextCall(options));
+    };
+  }
+}
+
+export class RetryUntilTimeoutInterceptor {
+  private readonly logger: MomentoLogger;
+  private readonly responseDataReceivedTimeoutMs: number;
+  private readonly overallRequestTimeoutMs: number;
+  private overallDeadline?: Date;
+
+  constructor(
+    loggerFactory: MomentoLoggerFactory,
+    overallRequestTimeoutMs: number,
+    responseDataReceivedTimeoutMs: number
+  ) {
+    this.logger = loggerFactory.getLogger(this);
+    this.responseDataReceivedTimeoutMs = responseDataReceivedTimeoutMs;
+    this.overallRequestTimeoutMs = overallRequestTimeoutMs;
+  }
+
+  public createTimeoutInterceptor(): Interceptor {
+    return (options, nextCall) => {
+      // Replace default timeout with the desired overall timeout
+      // on the first time through and set the first incremental timeout
+      if (this.overallDeadline === undefined) {
+        const newDate = new Date(Date.now());
+        newDate.setMilliseconds(
+          newDate.getMilliseconds() + this.overallRequestTimeoutMs
+        );
+        this.overallDeadline = newDate;
+
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.responseDataReceivedTimeoutMs
+        );
+        options.deadline = deadline;
+        this.logger.debug(`new deadline set to ${options.deadline.valueOf()}`);
+        return new InterceptingCall(nextCall(options));
+      }
+
+      const receivedDeadline = options.deadline?.valueOf() || 0;
+      this.logger.debug(
+        `intercepting call with options.deadline ${receivedDeadline}`
+      );
+
+      // Reset incremental deadline only if it has been reached
+      if (receivedDeadline < Date.now()) {
+        this.logger.debug('received deadline < current time, resetting');
+
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.responseDataReceivedTimeoutMs
+        );
+
+        if (deadline.valueOf() > this.overallDeadline.valueOf()) {
+          this.logger.debug(
+            'Unable to successfully retry request within client timeout, canceling request'
+          );
```

Comment:
An important consideration here is how close to the deadline do we need to be to not bother retrying. In the current implementation, that value is `responseDataReceivedTimeoutMs`:

Consider an overall deadline at time=1000ms, a `responseDataReceivedTimeoutMs` of 100ms, and that we're now at time=925ms. Then the request deadline is `925ms + 100ms = 1025ms`, which exceeds our overall deadline, so we will not retry. But maybe the request could succeed in less than 75ms?

We could consider some percentage of the data received timeout instead, like 50%. Up for discussion.

So we need to decide how close to the `overallDeadline` do we need to be to not even bother retrying. Right now if are within `responseDataReceivedTimeoutMs` of the `overallDeadline`, we will not retry.

Reply:
For sure! If we do a percent-based margin around that, would that have to become another configuration option? Or will we just hardcode that in for now?

Reply:
would like to avoid more config options if possible. could we not just check `Date.now() > this.overallDeadline`, and if it's not, then we set the next deadline to `min(this.overallDeadline, deadline + this.responseDataReceivedTimeoutMs)`?